fiftyone/core/collections.py

```diff
@@ -66,6 +70,83 @@ def registrar(func):
     return registrar
 
 
+class SuccessTimeout(Exception):
+    def __init__(self, seconds):
+        message = "Succeeded after %s seconds" % seconds
+        logging.info(message)
+        super().__init__(message)
+
+
+class GracefulTimeoutError(Exception):
+    def __init__(self, seconds):
+        message = "Graceful timeout after %s seconds" % seconds
+        logging.info(message)
+        super().__init__(message)
+
+
+def timeout(*, seconds=3, graceful=False, max_timeout=30):
+    """Abort the wrapped function call after the specified number of seconds
+    have elapsed."""
+    timeout.validate_func = None
+    timeout.time_count = 0
+
+    def _handle_timeout(signum, frame):
+        validate_func = timeout.validate_func
+        time_count = timeout.time_count + seconds
+        if callable(validate_func):
+            result = validate_func()
+            if result:
+                timeout.time_count = 0
+                raise SuccessTimeout(time_count)
+            elif time_count < max_timeout:
+                timeout.time_count = time_count
+                signal(SIGALRM, _handle_timeout)
+                alarm(seconds)
+            else:
+                timeout.time_count = 0
+                raise (
+                    GracefulTimeoutError(time_count)
+                    if graceful
+                    else TimeoutError(
+                        "Function call timed out after %s seconds" % time_count
+                    )
+                )
+        else:
+            raise (
+                GracefulTimeoutError(time_count)
+                if graceful
+                else TimeoutError(
+                    "Function call timed out after %s seconds" % time_count
+                )
+            )
+
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            dataset = args[0]
+            index_name = args[1] if len(args) > 1 else kwargs.get("index_name")
```

Comment:
We have an implementation for more robust retrieving index-in-progress information here:
https://github.com/voxel51/fiftyone/pull/4633